tesla_api/energy.py

```diff
@@ -50,10 +59,39 @@ async def get_battery_count(self):
         info = await self.get_energy_site_info()
         return int(info["battery_count"])
 
-
     async def get_energy_site_live_status(self):
         return await self._api_client.get('energy_sites/{}/live_status'.format(self._energy_site_id))
 
+    async def get_energy_site_calendar_history_data(self, kind: str = 'energy', period: str = 'day',
+                                                    end_date: [str, datetime] = None,
+                                                    time_zone: str = None) -> Optional[dict]:
+        """
+        Returns historical energy data
+        :param kind: [power, energy, self_consumption]
+        :param period: Amount of time to include in report. One of day, week, month, year, and lifetime.
+                    When kind is 'power', this parameter is ignored, and the period is always 'day'
+        :param end_date: ISO 8601 datetime, e.g. 2019-12-23T17:39:18.546Z. either in str or as datetime
+                    The response report interval ends on this datetime and starts at the beginning of the given period
+                    at 1:00 AM.  Note time cannot be 00:00. Defaults to the current time.
+        :param time_zone: IANA/Olsen time zone identifier, e.g. America/New_York. If set, and if the end_date is a
+                    'naive' datetime object (without timezone info) this field will be used to set the timezone
+                    for the end_date.  Otherwise it is ignored
+        """
+        params = {'kind': kind, 'period': period}
+        if end_date is not None:
+            if type(end_date) is datetime:
+                if end_date.hour == 0:
+                    # if the datetime object's time is 00:00 then the API returns nothing. We adjust by adding an hour
```

Comment:
Hmm, interesting. I don't have a power wall to test this out with, but does 24:00 work, or 00:01?

Reply:
24:00 generates an API error, anything other than 00:00:00 works.  However, the API response starts at 01:00:00 in the specified timezone (regardless of the time added to the date), so if you're passing a datetime without a time component (as you normally would, given that the API will always return the data for one full day, starting at 1am on that date) it seems appropriate to add one hour if hours==0, as that is most likely the intent.  So, I'm keeping it as is